File: LEET_1944.py
class Solution:
    def canSeePersonsCount(self, heights):
        answer = [0] * len(heights)
        stack = []
        for i, height in enumerate(heights):
            while stack and heights[stack[-1]] <= height:
                # 새로운 height보다 키작은 이전 애들은 더이상 볼 수 없음
                answer[stack.pop()] += 1
            if stack:
                # 새로운 애들보다 키가 큰 첫번째 애도 볼 수 있음
                answer[stack[-1]] += 1
            stack.append(i)

File: LEET_1944.cpp
#include <vector>
#include <stack>

using namespace std;
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {
        int n = heights.size();
        vector<int> ans(n);
        stack<int> st;
        for (int i = n-1; i >= 0; i--) { // 뒤에서
            while (!st.empty() && heights[i] > st.top()) { // 오른쪽에 있는 작은 애들 봄
                st.pop();
                ans[i]++; // [i+1,] 에서 더 작은 것들
            }
            if (!st.empty())
                ans[i]++; // 현재 있는 것이 최장신이 아님 == 최장신을 볼 수 있음 
            st.push(heights[i]);
        }
        return ans;
    }
};